### DISCUSSION TITLE
Help with LPSPI DMA in Synchronous Mode on RT1060EVKB

### DISCUSSION THREAD

**Discussion by ofirshe** [zephyrproject-rtos/zephyr#81007] [Q&A]
Hi @DerekSnell and @decsny,

After rebasing to align with the latest fix in the LPSPI driver [Fix NXP LPSPI native chip select when using synchronous API with DMA](https://github.com/zephyrproject-rtos/zephyr/pull/78732), I’m seeing some unexpected behavior. For reference, I’m working with the RT1060EVKB board and have the following project configuration flags:
```
CONFIG_SPI=y
CONFIG_SPI_MCUX_LPSPI_DMA=y
CONFIG_SPI_ASYNC=n
```
I also updated the device tree to enable SPI transactions with DMA on the LPSPI peripheral:
```
&lpspi1 {
	status = "okay";
	/* DMA channels 0 and 1, muxed to LPSPI1 RX and TX */
	dmas = <&edma0 0 13>, <&edma0 1 14>;
	dma-names = "rx", "tx";
	pinctrl-0 = <&pinmux_lpspi1>;
	pinctrl-names = "default";
	transfer-delay = <500>;
	si5402: spi-dev-si5402@0 {
		compatible = "skyworks,si5402";
		status = "okay";
		spi-max-frequency = <1000000>;
		reg = <0>;
	};
};
```

Here’s the code I’m using for the transaction:

```
uint8_t tx_data[2] = { 0 };
uint8_t rx_data[4]  = { 0 };
uint16_t rx_len;
uint16_t tx_len;

tx_data[0] = 0xD0;
tx_len = 1;
rx_len = 3;
const struct spi_buf tx_buf = { .buf = tx_data, .len = tx_len };
const struct spi_buf rx_buf = { .buf = rx_data, .len = rx_len };
const struct spi_buf_set tx = { .buffers = &tx_buf, .count = 1 };
const struct spi_buf_set rx = { .buffers = &rx_buf, .count = 1 };
return spi_transceive_dt(&cfg->bus, &tx, &rx);
```

SPI configured mode: `SPI_WORD_SET(8) | SPI_OP_MODE_MASTER | SPI_TRANSFER_MSB`

From the waveform output , I noticed some issues:
<img width="890" alt="image" src="https://github.com/user-attachments/assets/d5b42cc3-f9f2-4257-884b-72bcdce7cc62">

1. The clock line remains high throughout the transaction, which is unexpected, as I’ve configured it for 1 MHz.
2. The MOSI channel does not display the expected data (0xD0).

Additionally, `spi_transceive_dt` returns `-116`, with an error log indicating `<err> spi_mcux_lpspi: Timeout waiting for transfer complete.`

When using the default LPSPI mode (non-DMA path) with the `cs-gpios` workaround in the DTS, I noticed this output from the digital analyzer:

<img width="1240" alt="image" src="https://github.com/user-attachments/assets/b67eaa9f-bff8-406e-9480-89544c658eb7">

Aside from the issue with the component on the MISO channel (which I'm troubleshooting separately), I'm unsure which of the two modes reflects the correct behavior for the code I shared above. In the DMA path, I observe a single transaction from the RT1060, whereas in the non-DMA path, I see two transactions.

Best regards, 
Ofir.

**Comment by DerekSnell** [comment]
This is confirmed as an issue in the LPSPI driver, and this GitHub Issue will be used for tracking and further communication:
[LPSPI driver with DMA does not work on NXP RT platform](https://github.com/zephyrproject-rtos/zephyr/issues/81033)

**Reply by ofirshe** [reply]
Thanks @DerekSnell for the update. 

Could you confirm the expected behavior of the LPSPI in non-DMA mode (assuming it is functioning correctly)? Should the RT1060 be issuing two separate transactions for the code I provided?

BR

**Reply by decsny** [reply]
non DMA mode chip select also doesn't work correctly, this is longstanding issue https://github.com/zephyrproject-rtos/zephyr/issues/16544

**Reply by ofirshe** [reply]
@decsny I mentioned that I'm using the `cs-gpios` workaround in the DTS to address the CS issue, but my question is specifically about the expected driver behavior for the code I attached. I’m uncertain why it's sending two transactions instead of one.

**Comment by DerekSnell** [comment]
Linking to your newer discussion https://github.com/zephyrproject-rtos/zephyr/discussions/81264, as it seems to be a duplicate
